### ISSUE TITLE
curl has error code is 28

### ISSUE THREAD

**Issue by hk520570** [curl/curl#4052]
I use curl_easy_perform，if I use mutiple call it，it hasn't lock， return error code is 28，when i use lock it work well~
### operating system
Windows 7

**Comment by bagder** [comment]
Please provide an example code that reproduces this problem. What libcurl version are you using?

**Comment by hk520570** [comment]
> Please provide an example code that reproduces this problem. What libcurl version are you using?

  `  std::thread ths([=]
                    {
                        if(!FileUtils::getInstance()->isDirectoryExist(savepath))
                        {
                            FileUtils::getInstance()->createDirectory(savepath);
                        }
                        std::string savefile = savepath + savename;
                        if (savepath[savepath.length() - 1] != '/')
                        {
                            savefile = savepath + "/" + savename;
                        }
                        FILE *fp = fopen(savefile.c_str(), "wb");
                        int nRes = -1;
                        if(nullptr == fp)
                        {
                            std::string str = StringUtils::format("open file:%s error!", savefile.c_str());
                            safeNotifyWithRelease(nRes, str, savename);
                            return;
                        }
                        curl_easy_setopt(m_pCurl, CURLOPT_URL, m_strUrl.c_str());
                        curl_easy_setopt(m_pCurl, CURLOPT_WRITEFUNCTION, downLoadPackage);
                        //curl_easy_setopt(m_pCurl, CURLOPT_WRITEFUNCTION, ProcessData);
						curl_easy_setopt(m_pCurl, CURLOPT_WRITEDATA, fp);
                        curl_easy_setopt(m_pCurl, CURLOPT_NOPROGRESS, false);
                        curl_easy_setopt(m_pCurl, CURLOPT_PROGRESSFUNCTION, downloaderProgressFunc);
                        curl_easy_setopt(m_pCurl, CURLOPT_PROGRESSDATA, this);
                        curl_easy_setopt(m_pCurl, CURLOPT_NOSIGNAL, 1L);
                        curl_easy_setopt(m_pCurl, CURLOPT_CONNECTTIMEOUT, connecttimeout);
                        curl_easy_setopt(m_pCurl, CURLOPT_TIMEOUT,downloadtimeout);
                        curl_easy_setopt(m_pCurl, CURLOPT_LOW_SPEED_LIMIT, 1L);
                        curl_easy_setopt(m_pCurl, CURLOPT_LOW_SPEED_TIME, 5L);
                        curl_easy_setopt(m_pCurl, CURLOPT_FOLLOWLOCATION, 1 );
					
			curl_easy_setopt(m_pCurl, CURLOPT_SSL_VERIFYPEER, 0L);
			curl_easy_setopt(m_pCurl, CURLOPT_SSL_VERIFYHOST, 0L);
			curl_easy_setopt(m_pCurl, CURLOPT_VERBOSE, 0L);
			CURLcode res = curl_easy_perform(m_pCurl);
                        if (CURLE_OK == res)
                        {
                            double downsize = 0;
                            res = curl_easy_getinfo(m_pCurl, CURLINFO_CONTENT_LENGTH_DOWNLOAD, &downsize);
                            if (CURLE_OK == res && downsize > 0)
                            {
                                nRes = 0;
                            }
                        }
						else {
							nRes = res;
						}
						fclose(fp);
                    });
    ths.detach();`
curl_easy_perform is not work well(Mutiple call it)，some return error 28
version is 7.52.1

**Comment by TedLyngmo** [comment]
@hk520570  If I understand the code snipplet correctly, you're creating a detached thread using a copy of the `CURL` instance (`m_pCurl`)., right? You probably should `curl_easy_duphandle` in the beginning of the thread and end the thread with `curl_easy_cleanup` of the duplicated handle.

**Comment by hk520570** [comment]
I use it by Class，but my code when use it onece can release m_pCurl

**Comment by hk520570** [comment]
> @hk520570 If I understand the code snipplet correctly, you're creating a detached thread using a copy of the `CURL` instance (`m_pCurl`)., right? You probably should `curl_easy_duphandle` in the beginning of the thread and end the thread with `curl_easy_cleanup` of the duplicated handle.

I use it by Class，m_pCrul will be release（curl_easy_cleanup） onece call it

**Comment by TedLyngmo** [comment]
> I use it by Class，m_pCrul will be release（curl_easy_cleanup） onece call it

If you create a minimal example that can be used to reproduce the problem it'll be easier to understand. Is it possible that you are using the handle in multiple threads simultaneously? "_you must never use a single handle from more than one thread at any given time_", to quote the documentation.

**Comment by hk520570** [comment]
> > I use it by Class，m_pCrul will be release（curl_easy_cleanup） onece call it
> 
> If you create a minimal example that can be used to reproduce the problem it'll be easier to understand. Is it possible that you are using the handle in multiple threads simultaneously? "_you must never use a single handle from more than one thread at any given time_", to quote the documentation.

thank you，I try it

**Comment by jay** [comment]
No update. Probably user error. Feel free to reopen.
